pkg/chunk/aws/s3_storage_client.go

```diff
@@ -56,42 +76,122 @@ func (cfg *S3Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+		bucketNames:     bucketNames,
+		delimiter:       delimiter,
+		sseEncryption:   sseEncryption,
+		putUserMetadata: cfg.PutUserMetadata,
 	}
 	return &client, nil
 }
 
+func buildS3Config(cfg S3Config) (*aws.Config, []string, error) {
+	var s3Config *aws.Config
+	var err error
+
+	// if an s3 url is passed use it to initialize the s3Config and then override with any additional params
+	if cfg.S3.URL != nil {
+		s3Config, err = awscommon.ConfigFromURL(cfg.S3.URL)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		s3Config = &aws.Config{}
+		s3Config = s3Config.WithRegion("dummy")
+		s3Config = s3Config.WithCredentials(credentials.AnonymousCredentials)
+	}
+
+	s3Config = s3Config.WithMaxRetries(0) // We do our own retries, so we can monitor them
+	s3Config = s3Config.WithHTTPClient(&http.Client{Transport: defaultTransport})
+	s3Config = s3Config.WithS3ForcePathStyle(cfg.S3ForcePathStyle) // support for Path Style S3 url if has the flag
+
+	if cfg.Endpoint != "" {
+		s3Config = s3Config.WithEndpoint(cfg.Endpoint)
+	}
+
+	if cfg.Insecure {
+		s3Config = s3Config.WithDisableSSL(true)
+	}
+
+	if cfg.Region != "" {
+		s3Config = s3Config.WithRegion(cfg.Region)
+	}
+
+	if cfg.AccessKeyID != "" && cfg.SecretAccessKey == "" ||
+		cfg.AccessKeyID == "" && cfg.SecretAccessKey != "" {
+		return nil, nil, errors.New("Failed to initialize s3_storage_client:  Must supply both an Access Key ID and Secret Access Key or neither")
+	}
+
+	if cfg.AccessKeyID != "" && cfg.SecretAccessKey != "" {
+		creds := credentials.NewStaticCredentials(cfg.AccessKeyID, cfg.SecretAccessKey, "")
+		s3Config = s3Config.WithCredentials(creds)
+	}
+
+	// Use a custom http.Client with the golang defaults but also specifying
+	// MaxIdleConnsPerHost because of a bug in golang https://github.com/golang/go/issues/13801
+	// where MaxIdleConnsPerHost does not work as expected.
+	s3Config = s3Config.WithHTTPClient(&http.Client{
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}).DialContext,
+			MaxIdleConns:          100,
+			IdleConnTimeout:       cfg.HTTPConfig.IdleConnTimeout, //90 * time.Second,
+			MaxIdleConnsPerHost:   100,
+			TLSHandshakeTimeout:   3 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+			ResponseHeaderTimeout: time.Duration(cfg.HTTPConfig.ResponseHeaderTimeout),
+			TLSClientConfig:       &tls.Config{InsecureSkipVerify: cfg.HTTPConfig.InsecureSkipVerify},
+		},
+	})
+
+	// bucketnames
+	var bucketNames []string
+	if cfg.S3.URL != nil {
+		bucketNames = []string{strings.TrimPrefix(cfg.S3.URL.Path, "/")}
+	}
+
+	if cfg.BucketNames != "" {
+		bucketNames = strings.Split(cfg.BucketNames, ",") // comma separated list of bucket names
+	}
+
+	if len(bucketNames) == 0 {
+		return nil, nil, errors.New("Failed to initialize s3_storage_client:  At least one bucket name must be specified")
```

Comment:
When returning errors in this function, I would just include the second half of the error and wrap/return the first half in the calling function.